File: acgs2-core/sdk/typescript/src/services/hitl-approvals.ts
/**
 * ACGS-2 HITL Approvals Service
 * Human-in-the-Loop approval workflows
 * Constitutional Hash: cdd01ef066bc6cf2
 */

import type { ACGS2Client } from '../client/index.js';
import { CONSTITUTIONAL_HASH } from '../types/index.js';
import type {
  ApprovalRequest,
  CreateApprovalRequest,
  PaginatedResponse,
  SubmitApprovalDecision,
} from '../types/index.js';

export class HITLApprovalsService {
  private client: ACGS2Client;
  private basePath = '/api/v1/hitl-approvals';

  constructor(client: ACGS2Client) {
    this.client = client;
  }

  async createApprovalRequest(request: CreateApprovalRequest): Promise<ApprovalRequest> {
    const response = await this.client.post(`${this.basePath}/approvals`, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as ApprovalRequest;
  }

  async getApprovalRequest(requestId: string): Promise<ApprovalRequest> {
    const response = await this.client.get(`${this.basePath}/approvals/${requestId}`);
    return response.data as ApprovalRequest;
  }

  async listApprovalRequests(params?: {
    page?: number;
    pageSize?: number;
    status?: string;
    requesterId?: string;
    pendingFor?: string;
  }): Promise<PaginatedResponse<ApprovalRequest>> {
    const queryParams = new URLSearchParams();
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.pageSize) queryParams.append('pageSize', params.pageSize.toString());
    if (params?.status) queryParams.append('status', params.status);
    if (params?.requesterId) queryParams.append('requesterId', params.requesterId);
    if (params?.pendingFor) queryParams.append('pendingFor', params.pendingFor);

    const response = await this.client.get(
      `${this.basePath}/approvals?${queryParams.toString()}`
    );
    return response.data as PaginatedResponse<ApprovalRequest>;
  }

  async submitDecision(
    requestId: string,
    decision: SubmitApprovalDecision
  ): Promise<ApprovalRequest> {
    const response = await this.client.post(`${this.basePath}/approvals/${requestId}/decisions`, {
      ...decision,
      timestamp: new Date().toISOString(),
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as ApprovalRequest;
  }

  async escalateApprovalRequest(requestId: string, reason: string): Promise<ApprovalRequest> {
    const response = await this.client.post(`${this.basePath}/approvals/${requestId}/escalate`, {
      reason,
    });
    return response.data as ApprovalRequest;
  }

  async cancelApprovalRequest(requestId: string, reason?: string): Promise<void> {
    await this.client.post(`${this.basePath}/approvals/${requestId}/cancel`, {
      reason,
    });
  }

  async getPendingApprovals(
    userId: string,
    params?: { page?: number; pageSize?: number }
  ): Promise<PaginatedResponse<ApprovalRequest>> {
    const queryParams = new URLSearchParams();
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.pageSize) queryParams.append('pageSize', params.pageSize.toString());

    const response = await this.client.get(
      `${this.basePath}/approvals/pending/${userId}?${queryParams.toString()}`
    );
    return response.data as PaginatedResponse<ApprovalRequest>;
  }

  async getApprovalWorkflowConfig(): Promise<Record<string, unknown>> {
    const response = await this.client.get(`${this.basePath}/config`);
    return response.data as Record<string, unknown>;
  }

  async updateApprovalWorkflowConfig(config: Record<string, unknown>): Promise<Record<string, unknown>> {
    const response = await this.client.put(`${this.basePath}/config`, {
      config,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
    return response.data as Record<string, unknown>;
  }

  async getApprovalMetrics(params?: {
    startDate?: string;
    endDate?: string;
  }): Promise<Record<string, unknown>> {
    const queryParams = new URLSearchParams();
    if (params?.startDate) queryParams.append('startDate', params.startDate);
    if (params?.endDate) queryParams.append('endDate', params.endDate);

    const response = await this.client.get(
      `${this.basePath}/metrics?${queryParams.toString()}`
    );
    return response.data as Record<string, unknown>;
  }
}

File: src/core/sdk/python/acgs2_sdk/services/hitl_approvals.py
"""
ACGS-2 HITL Approvals Service
Human-in-the-Loop approval workflows
Constitutional Hash: cdd01ef066bc6cf2
"""

from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any, Union

try:
    from src.core.shared.types import JSONDict, JSONValue
except ImportError:
    JSONValue = Union[str, int, float, bool, None, dict[str, Any], list[Any]]  # type: ignore[misc]
    JSONDict = dict[str, JSONValue]  # type: ignore[misc]

from acgs2_sdk.constants import CONSTITUTIONAL_HASH, HITL_APPROVALS_ENDPOINT
from acgs2_sdk.models import (
    ApprovalRequest,
    ApprovalStatus,
    CreateApprovalRequest,
    PaginatedResponse,
    SubmitApprovalDecision,
)

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class HITLApprovalsService:
    """Service for Human-in-the-Loop approval workflows."""

    def __init__(self, client: "ACGS2Client") -> None:
        self._client = client
        self._base_path = HITL_APPROVALS_ENDPOINT

    async def create_approval_request(
        self,
        request: CreateApprovalRequest,
    ) -> ApprovalRequest:
        """Create an approval request."""
        data = await self._client.post(
            f"{self._base_path}/approvals",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return ApprovalRequest.model_validate(data.get("data", data))

    async def get_approval_request(self, request_id: str) -> ApprovalRequest:
        """Get an approval request by ID."""
        data = await self._client.get(f"{self._base_path}/approvals/{request_id}")
        return ApprovalRequest.model_validate(data.get("data", data))

    async def list_approval_requests(
        self,
        page: int = 1,
        page_size: int = 50,
        status: ApprovalStatus | None = None,
        requester_id: str | None = None,
        pending_for: str | None = None,
    ) -> PaginatedResponse[ApprovalRequest]:
        """List approval requests."""
        params: JSONDict = {"page": page, "pageSize": page_size}
        if status:
            params["status"] = status.value
        if requester_id:
            params["requesterId"] = requester_id
        if pending_for:
            params["pendingFor"] = pending_for

        data = await self._client.get(f"{self._base_path}/approvals", params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[ApprovalRequest](
            data=[ApprovalRequest.model_validate(r) for r in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def submit_decision(
        self,
        request_id: str,
        decision: SubmitApprovalDecision,
    ) -> ApprovalRequest:
        """Submit an approval decision."""
        data = await self._client.post(
            f"{self._base_path}/approvals/{request_id}/decisions",
            json={
                **decision.model_dump(by_alias=True),
                "timestamp": datetime.now(UTC).isoformat(),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return ApprovalRequest.model_validate(data.get("data", data))

    async def escalate(
        self,
        request_id: str,
        reason: str,
    ) -> ApprovalRequest:
        """Escalate an approval request."""
        data = await self._client.post(
            f"{self._base_path}/approvals/{request_id}/escalate",
            json={"reason": reason},
        )
        return ApprovalRequest.model_validate(data.get("data", data))

    async def cancel_approval_request(
        self,
        request_id: str,
        reason: str | None = None,
    ) -> None:
        """Cancel an approval request."""
        await self._client.post(
            f"{self._base_path}/approvals/{request_id}/cancel",
            json={"reason": reason},
        )

    async def get_pending_approvals(
        self,
        user_id: str,
        page: int = 1,
        page_size: int = 50,
    ) -> PaginatedResponse[ApprovalRequest]:
        """Get pending approvals for a user."""
        params = {"page": page, "pageSize": page_size}
        data = await self._client.get(
            f"{self._base_path}/approvals/pending/{user_id}", params=params
        )
        response_data = data.get("data", data)
        return PaginatedResponse[ApprovalRequest](
            data=[ApprovalRequest.model_validate(r) for r in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def get_approval_workflow_config(self) -> JSONDict:
        """Get approval workflow configuration."""
        data = await self._client.get(f"{self._base_path}/config")
        return data.get("data", data)

    async def update_approval_workflow_config(
        self,
        config: JSONDict,
    ) -> JSONDict:
        """Update approval workflow configuration."""
        data = await self._client.put(
            f"{self._base_path}/config",
            json={
                "config": config,
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return data.get("data", data)

    async def get_approval_metrics(
        self,
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> JSONDict:
        """Get approval workflow metrics."""
        params: JSONDict = {}
        if start_date:
            params["startDate"] = start_date
        if end_date:
            params["endDate"] = end_date

        data = await self._client.get(f"{self._base_path}/metrics", params=params)
        return data.get("data", data)
